streampark-console/streampark-console-service/src/main/scala/org/apache/streampark/console/core/task/FlinkK8sChangeListener.scala

```diff
@@ -0,0 +1,222 @@
+                    trackedKey.clusterNamespace == namespaceAndName._1 && trackedKey.clusterName == namespaceAndName._2)
+                .someOrFail(TrackKeyNotFound(namespaceAndName._1, namespaceAndName._2))
+              val restSvcEndpoint: RestSvcEndpoint = restSvcEndpointSnap._2
+
+              val app: Application = applicationService.getById(trackKey.map(_.id))
+
+              val flinkCluster: FlinkCluster = flinkClusterService.getById(app.getFlinkClusterId)
+
+              if (restSvcEndpoint == null || restSvcEndpoint.chooseRest == null) return ZIO.unit
+              val url = restSvcEndpoint.chooseRest
+              app.setFlinkRestUrl(url)
+              applicationService.persistMetrics(app)
+
+              flinkCluster.setAddress(url)
+              flinkClusterService.update(flinkCluster)
+
+            }
+            .retryN(3)
+            .ignore
+      }
+  }
+
+  private def setByJobStatusCV(app: Application, jobSnapshot: JobSnapshot): Unit = { // infer the final flink job state
+    val evalJobState: EvalJobState = inferEvalJobStateFromPersist(
+      jobSnapshot.evalState,
+      FlinkAppStateConverter.flinkAppStateToK8sEvalJobState(FlinkAppState.of(app.getState)))
+
+    val state: FlinkAppState = FlinkAppStateConverter.k8sEvalJobStateToFlinkAppState(evalJobState)
+    val jobStatusOption: Option[JobStatus] = jobSnapshot.jobStatus
+
+    if (jobStatusOption.nonEmpty) {
+      val jobStatus: JobStatus = jobStatusOption.get
+      // corrective start-time / end-time / duration
+      val preStartTime: Long =
+        if (app.getStartTime != null) app.getStartTime.getTime
+        else 0
+
+      val startTime: Long = Math.max(jobStatus.startTs, preStartTime)
+      val preEndTime: Long =
+        if (app.getEndTime != null) app.getEndTime.getTime
+        else 0
+      var endTime: Long = Math.max(jobStatus.endTs.getOrElse(-1L), preEndTime)
+      var duration: Long = if (app.getDuration != null) app.getDuration else 0
+      if (FlinkAppState.isEndState(state.getValue)) {
+        if (endTime < startTime) endTime = System.currentTimeMillis
+        if (duration <= 0) duration = endTime - startTime
+      }
+      app.setJobId(jobStatus.jobId)
+      val totalTask = if (jobStatus.tasks.nonEmpty) jobStatus.tasks.get.total else 0
+      app.setTotalTask(totalTask)
+      app.setStartTime(
+        new Date(
+          if (startTime > 0) startTime
+          else 0))
+      app.setEndTime(
+        if (endTime > 0 && endTime >= startTime) new Date(endTime)
+        else null)
+      app.setDuration(
+        if (duration > 0) duration
+        else 0)
+    }
+
+    app.setState(state.getValue)
+    // when a flink job status change event can be received, it means
+    // that the operation command sent by streampark has been completed.
+    app.setOptionState(OptionState.NONE.getValue)
+  }
+
+  def inferEvalJobStateFromPersist(current: EvalJobState, previous: EvalJobState): EvalJobState = {
```

Comment:
No need to infer the current job status from persistent storage, directly use the job status obtained from the FlinkK8sObserver.evaluatedJobSnaps subscription.